### ISSUE TITLE
php: unable to load remote files within a docker alpine image

### ISSUE THREAD

**Issue by michabbb** [Imagick/imagick#322]
hi there, i don´t know if this is the right place to address my issue, so if not, please let me know where.
the problem is, this line of code:

`new \Imagick('https://www.chip.de/ii/3/0/2/8/5/eb3b8af79bd79970.jpg');`

throws this exception: 

```
ImagickException: no decode delegate for this image format `' @ error/constitute.c/ReadImage/562
```

when I test this with a debian image, there is no error, so I am very sure, this has something todo with alpine. i don´t know if there is something wrong inside alpine or imagick, no idea 😕 

here`s a simple Dockerfile to reproduce this:

```
FROM php:7-fpm-alpine

COPY docker-php-ext-* docker-php-entrypoint /usr/local/bin/
RUN chmod u+x  /usr/local/bin/docker-php-ext-* /usr/local/bin/docker-php-entrypoint

# imagick
RUN apk add --update --no-cache autoconf g++ imagemagick-dev libtool libxml2-dev curl strace make pcre-dev imagemagick libstdc++ libpng-dev libjpeg-turbo-dev zlib-dev libtool libbz2 \
    bzip2-dev \
    libstdc++ \
    libxslt-dev \
    imagemagick-dev \
    && pecl install imagick \
    && docker-php-ext-enable imagick \
    && apk del autoconf g++ libtool make pcre-dev
```
and [here](https://gist.github.com/michabbb/5107a401f9384f5dcbdeab3fd15c3d56) is the `phpinfo()` output of my size optimized image `michabbb/php7:slim`:

```
PHP Version => 7.4.3
imagick module => enabled
imagick module version => 3.4.4
imagick classes => Imagick, ImagickDraw, ImagickPixel, ImagickPixelIterator, ImagickKernel
Imagick compiled with ImageMagick version => ImageMagick 7.0.10-0 Q16 x86_64 2020-03-08 https://imagemagick.org
Imagick using ImageMagick library version => ImageMagick 7.0.10-0 Q16 x86_64 2020-03-08 https://imagemagick.org
ImageMagick copyright => © 1999-2020 ImageMagick Studio LLC
ImageMagick release date => 2020-03-08
ImageMagick number of supported formats:  => 214
ImageMagick supported formats => 3FR, 3G2, 3GP, AAI, ART, ARW, AVI, AVS, BGR, BGRA, BGRO, BMP, BMP2, BMP3, BRF, CAL, CALS, CANVAS, CAPTION, CIN, CIP, CLIP, CMYK, CMYKA, CR2, CR3, CRW, CUBE, CUR, CUT, DATA, DCM, DCR, DCRAW, DCX, DDS, DFONT, DNG, DOT, DPX, DXT1, DXT5, EPS2, EPS3, EPT, EPT2, EPT3, ERF, FAX, FILE, FITS, FLV, FRACTAL, FTP, FTS, G3, G4, GIF, GIF87, GRADIENT, GRAY, GRAYA, GV, HALD, HDR, HISTOGRAM, HRZ, HTM, HTML, HTTP, HTTPS, ICB, ICO, ICON, IIQ, INFO, INLINE, IPL, ISOBRL, ISOBRL6, JNG, JNX, JPE, JPEG, JPG, JPS, JSON, K25, KDC, LABEL, M2V, M4V, MAC, MAP, MASK, MAT, MATTE, MEF, MIFF, MKV, MNG, MONO, MOV, MP4, MPC, MPEG, MPG, MRW, MSL, MTV, MVG, NEF, NRW, NULL, ORF, OTB, OTF, PAL, PALM, PAM, PANGO, PATTERN, PBM, PCD, PCDS, PCL, PCT, PCX, PDB, PEF, PES, PFA, PFB, PFM, PGM, PGX, PICON, PICT, PIX, PJPEG, PLASMA, PNG, PNG00, PNG24, PNG32, PNG48, PNG64, PNG8, PNM, PPM, PS2, PS3, PSB, PSD, PWP, RADIAL-GRADIENT, RAF, RAS, RAW, RGB, RGB565, RGBA, RGBO, RGF, RLA, RLE, RMF, RW2, SCR, SCT, SFW, SGI, SHTML, SIX, SIXEL, SPARSE-COLOR, SR2, SRF, STEGANO, SUN, TEXT, TGA, THUMBNAIL, TILE, TIM, TM2, TTC, TTF, TXT, UBRL, UBRL6, UIL, UYVY, VDA, VICAR, VID, VIFF, VIPS, VST, WBMP, WMV, WPG, X, X3F, XBM, XC, XCF, XPM, XPS, XV, XWD, YCbCr, YCbCrA, YUV

Directive => Local Value => Master Value
imagick.locale_fix => 0 => 0
imagick.progress_monitor => 0 => 0
imagick.skip_version_check => 0 => 0
```

to sum this up: loading a remote file with Debian works, with alpine not.

**Comment by Danack** [comment]
Just to sanity check, do you get any warnings from PHP when it starts up?

> imagick.skip_version_check

**Comment by Danack** [comment]
Oh, 'kay.

```
new \Imagick('https://www.chip.de/ii/3/0/2/8/5/eb3b8af79bd79970.jpg');
```
I strongly recommend not doing this. Download the file first, and then open it locally after doing some sanity checks on the file type with the code in https://github.com/imagick/imagick#security

It's very likely the download is failing for any number of reasons:

* You don't have the right permissions for the process.
* You don't have the right settings in the ImageMagick policy file.
* The server on the other end doesn't like the user-agent used by ImageMagick, or maybe the lack of a user-agent.
* The server is giving back a subtly invalid image, that the browsers know how to work around.

etc...

Downloading the file first allows you to inspect the file first, and can be more secure.

I've actually been thinking about removing this functionality but providing some simple code to do the download and check first.

If you wanted to investigate this, the way to do it would be to run the script through strace on the platform it's not working on. That will give a list of all the system calls being made. There will almost certainly be some failing calls near the end, that will indicate where the failure is originating. 

If you posted that output of strace here (though I might also need a comparison strace from debian where it's working) I can inspect it for the problem.

**Comment by michabbb** [comment]
@Danack  before i post all the debugs: thanks for your feedback and the hints.
thanks to this issue, I realized that fetching the image on my own is "of course" a better idea, I totally agree. the code is very old and because I was lazy that times, I let imagick handle the fetching, easy job, no doubt 😏 in my situation it´s not a problem, the script is not accessible to anybody, its a background job and I have full control over the image-server, where I get my images from. the URL here in my first posting is only an example - so from the security point of view, this is not a deal at all, but I am thanks for mentioning this. 

> do you get any warnings from PHP when it starts up?

no: not a single warning, everything seems to be absolutely fine!
working with local images works, no problem.

this is the script I used for testing:

```php
<?php

$tmpImg = new \Imagick('https://www.chip.de/ii/3/0/2/8/5/eb3b8af79bd79970.jpg');
print_r($tmpImg);
```

**so here´s a env with debian**, where using a URL is working:

```
FROM php:7.4.3-fpm

ADD https://raw.githubusercontent.com/mlocati/docker-php-extension-installer/master/install-php-extensions /usr/local/bin/
RUN chmod uga+x /usr/local/bin/install-php-extensions && sync 

RUN apt-get update && apt-get install -y strace
RUN install-php-extensions imagick

ENTRYPOINT ["docker-php-entrypoint"]

RUN sed -i -e 's/listen.*/listen = 0.0.0.0:9000/' /usr/local/etc/php-fpm.conf

RUN echo "expose_php=0" > /usr/local/etc/php/php.ini

CMD ["php-fpm"]
```

this is how I run the script **with debian**:

`docker run -v $(pwd):/app -w /app  --rm php:7.4-fpm-test3 strace -o /app/strace.debug.debian.txt php test.php`

[here](https://gist.github.com/michabbb/ea4ebb73e3a58bcf86f082070871a5a7) is the **strace with debian**.
[here](https://gist.github.com/michabbb/2ae20659d25ba446dc8880a4d92be08d) **is the strace of alpine**.

while doing all the tests, I noticed, this difference:

debian: `Imagick compiled with ImageMagick version => ImageMagick 6.9.10-23 Q16 x86_64`
alpine: `Imagick compiled with ImageMagick version => ImageMagick 7.0.9-7 Q16 x86_64`

for the alpine test env, let me mention:

- the URL is okay and accessible (even via https)
- this code works with alpine totally fine, so the system itself is able to fetch remote files:

```php
<?php
$remot_file = file_get_contents('https://www.chip.de/ii/3/0/2/8/5/eb3b8af79bd79970.jpg');
file_put_contents('/app/test.jpg',$remot_file);
````

I am totally fine "getting remote files on my own" - but after investing so much time into this issue, I am curious what the reason is and I happy to provide you with everything you need if you are interested to find the cause for this 😏 💪 

cheers,
micha

**Comment by Danack** [comment]
Okay, I can see the error just from the Alpine log. I'm going to note how I debugged it for fund-raising purposes*.

Initially, I opened the trace, and started reading from the start.....then I saw how long it was and started skimming it until I could see some interesting stuff.

The interesting stuff I picked out, is the bit where there start to be lots of read system calls in a row e.g. L1018-L1500 These are quite typical of network or disk activity.

Scrolling back up a bit, we find the a good place to get stuck into the code.

The following are snippets that I i) understand and ii) think are relevant. There are other bits that I don't understand, and bits that are probably not relevant.


### L943 - Attempt to open the file locally, then try to open it over http.
```
stat("https://www.chip.de/ii/3/0/2/8/5/eb3b8af79bd79970.jpg", 0x7ffeb41a3d90) = -1 ENOENT (No such file or directory)
stat("https://www.chip.de/ii/3/0/2/8/5/eb3b8af79bd79970.jpg", 0x7ffeb41a3d90) = -1 ENOENT (No such file or directory)

 Opening it locally failed, so try http.

socket(AF_INET6, SOCK_DGRAM, IPPROTO_IP) = 3
close(3)                                = 0
open("/etc/hosts", O_RDONLY|O_CLOEXEC)  = 3
fcntl(3, F_SETFD, FD_CLOEXEC)           = 0
read(3, "127.0.0.1\tlocalhost\n::1\tlocalhos"..., 1024) = 174
read(3, "", 1024)                       = 0
close(3)                                = 0
open("/etc/resolv.conf", O_RDONLY|O_CLOEXEC) = 3
fcntl(3, F_SETFD, FD_CLOEXEC)           = 0
read(3, "domain localdomain\nsearch locald"..., 248) = 63
read(3, "", 248)                        = 0
close(3)                                = 0
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 3
bind(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("0.0.0.0")}, 16) = 0
sendto(3, "CD\1\0\0\1\0\0\0\0\0\0\3www\4chip\2de\0\0\1\0\1", 29, MSG_NOSIGNAL, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.235.2")}, 16) = 29
// many lines of domain name lookup here.
```
So it looks like the http request is just about to start.


### L957 - Open connection to remote computer and download the file. The sendto it obviously the http request headers being sent, and then the reads look like the response being read.
```
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 3
bind(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("0.0.0.0")}, 16) = 0
sendto(3, "CD\1\0\0\1\0\0\0\0\0\0\3www\4chip\2de\0\0\1\0\1", 29, MSG_NOSIGNAL, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.235.2")}, 16) = 29
sendto(3, "C\372\1\0\0\1\0\0\0\0\0\0\3www\4chip\2de\0\0\34\0\1", 29, MSG_NOSIGNAL, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.235.2")}, 16) = 29
poll([{fd=3, events=POLLIN}], 1, 2500)  = 1 ([{fd=3, revents=POLLIN}])
recvfrom(3, "CD\201\200\0\1\0\4\0\0\0\0\3www\4chip\2de\0\0\1\0\1\300\f\0"..., 512, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.235.2")}, [16]) = 141
recvfrom(3, "C\372\201\200\0\1\0\4\0\0\0\0\3www\4chip\2de\0\0\34\0\1\300\f\0"..., 512, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.235.2")}, [16]) = 165
... snip
... snip
... snip
getgid()                                = 0
getegid()                               = 0
open("/etc/ssl/cert.pem", O_RDONLY)     = 4
brk(0x55b2ded49000)                     = 0x55b2ded49000
read(4, "-----BEGIN CERTIFICATE-----\nMIIH"..., 1024) = 1024
read(4, "xFwYIRKQ26ZIMApcQrAZQIDAQABo4ICy"..., 1024) = 1024
brk(0x55b2ded4a000)                     = 0x55b2ded4a000
read(4, "gIBAJcxAp/n/UNnSEQU5CmH7UwoZtCPN"..., 1024) = 1024
brk(0x55b2ded4b000)                     = 0x55b2ded4b000
read(4, "AiIwDQYJKoZIhvcNAQEBBQADggIPADCC"..., 1024) = 1024
brk(0x55b2ded4c000)                     = 0x55b2ded4c000
read(4, "8B1\nRXxlDPiyN8+sD8+Nb/kZ94/sHvJw"..., 1024) = 1024
read(4, "eQWN0YWxpcyBBdXRoZW50aWNhdGlvbiB"..., 1024) = 1024
... snip
... snip


So it looks like the data is transferred okay and we probably have the data on the local computer....'somewhere'.



### L1665 - open random on the computer to get some entropy to generate a random file name

```
open("/dev/urandom", O_RDONLY)          = 4
read(4, "!\346\224I\373\261h\224^\37\212\307\331\vX@[\266Sq\213\361\f\3\237\304\205\356'\243o\264"..., 64) = 64
close(4)                                = 0
getpid()  
```
There's probably a clone of the temp download to the unique name here, but I can't see it.

### L1680 - L1689 ImageMagick tries to open to local file and read from it....
```
stat("/tmp/magick-9jCypy-kHFiEA", {st_mode=S_IFREG|0600, st_size=0, ...}) = 0
open("/tmp/magick-9jCypy-kHFiEA", O_RDONLY) = 4
readv(4, [{iov_base="", iov_len=2}, {iov_base="", iov_len=1024}], 2) = 0
lseek(4, 0, SEEK_CUR)                   = 0
fstat(4, {st_mode=S_IFREG|0600, st_size=0, ...}) = 0
readv(4, [{iov_base="", iov_len=2051}, {iov_base="", iov_len=1024}], 2) = 0
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
fstat(4, {st_mode=S_IFREG|0600, st_size=0, ...}) = 0
close(4)   
```

From http://man7.org/linux/man-pages/man2/readv.2.html, which is found by googling ["linux system call readv"](https://www.google.com/search?q=linux+system+call+readv)

> On success, readv(), preadv() and preadv2() return the number of
>  bytes read; writev(), pwritev() and pwritev2() return the number of
>  bytes written.


And to pick out the two individual lines that have system calls on them:

readv(4, [{iov_base="", iov_len=2}, {iov_base="", iov_len=1024}], 2) = 0
readv(4, [{iov_base="", iov_len=2051}, {iov_base="", iov_len=1024}], 2) = 0

Okay, so....the data was downloaded, but when ImageMagick opens the tmp file, there is nothing there.


Okay, to the GOOGLE-MOBILE!!

Searching for "alpine temp file unreadable" gives nothing useful....kind of looks related to building alpine containers, not problems running them.

Searching for "alpine tmp file unreadable" gives....a very similar bug: https://github.com/spatie/laravel-medialibrary/issues/1613

> it's probably related to some permission setting in your environment. If there is something that can be done in medialibrary to fix this, send a PR.

Which.....smells correct. 

I'll now google "alpine tmp permissions"

The first link looks interesting.... [Alpine /tmp/nginx folder permissions](https://github.com/nginxinc/docker-nginx/issues/227) so I open that in a new tab to read in a moment.

But before I do, I see the seconds link looks even better: [packages that set wrong permissions of /var/tmp](https://gitlab.alpinelinux.org/alpine/aports/issues/9364)

> There are a few packages that change the permission of /var/tmp:
> nginx
> roundcubemail-openrc
> nextcloud-initscript
> Permission should be world writable (1777), but the above packages
> changes it to 0755.

So....I'm going to pass this back to you now as it looks like a permissions thing. 

If you follow that bug report and similar issues, you should be able to resolve it. Please let me know if it is that....If nothing else, I've been thinking about adding a debugging section along the lines of "If you see this error, then it's probably this problem". But I should probably fill that with actual issues, and not just my guesses. 

btw You might want to just write a simple php script that attempts to open a file in /tmp and see if that works, as that would be an easier thing to test than opening images over http.

**Comment by Danack** [comment]
> i don´t know if this is the right place to address my issue, so if not, please let me know where

This is one of the things I'm struggling with.

There isn't really a better place to ask for help, and at the same time, bug reports like this are (probably) not my fault....they're going to be the result of a really complex set of interactions, but not related to the stuff I look after.

**Comment by michabbb** [comment]
thank you very much for your investigation-story 😉 for me it´s a bit difficult to see this permission problem, because in my production system I always do `-v /tmp/:/tmp/` - so I use the tmp folder of my host - and even i don´t mount the `/tmp` of the host... a `ls -al /tmp` inside the container shows me:

`drwxrwxrwt    1 root     root          4096 Mar 13 18:17 tmp`

and i don´t see any chance to give this folder more rights 😕 
i am not a alpine expert here, but what i read from your links and compared to the "rights" the /tmp folder has, i don´t see any connection to these bug reports. but anyway, i will handle the "fetch image" stuff anyway, in a different way. so for me, it´s okay if you want to close this issue 😕

**Comment by Danack** [comment]
> I always do -v /tmp/:/tmp/

:spocks_eyebrow.gif:

That sounds like an exciting choice. 

You might want to try turning that off and seeing if the issue goes away. There's multiple settings on both the host and whatever is running your containers that could make files that have recently been created in the tmp system behave 'surprisingly'.

And yeah, I'm going to close this then as the issue is outside of anything I can support.

**Comment by michabbb** [comment]
it does not play a role if I use the `/tmp` of the host or not..... but anyway... a strange mystery 🧙‍♂️
thanks a lot for your support!! at the end, something positive came out: fetching files by myself is always better and gives me more control 😏 🎉
